recruiter/tests.py

```diff
@@ -46,20 +83,58 @@ def test_get_all_recruiters_of_a_specified_company_returns_all_recruiters_as_a_q
 
 @pytest.mark.django_db
 def test_if_user_is_a_recruiter_returns_true_if_the_auth_user_is_associated_with_a_recruiter_account(example_recruiter):
-    example_user_not_recruiter = User.objects.create_user("notrecruiter", "password")
+    example_user_not_recruiter = User.objects.create_user(
+        "notrecruiter", "password")
     example_user_not_recruiter.save()
     assert Recruiter.is_recruiter(example_user_not_recruiter.id) is False
     assert Recruiter.is_recruiter(example_recruiter.user.id) is True
 
 
 @pytest.mark.django_db
 def test_get_user_as_recruiter_object(example_recruiter):
-    assert isinstance(Recruiter.get_recruiter(example_recruiter.user.id), Recruiter)
+    assert isinstance(Recruiter.get_recruiter(
+        example_recruiter.user.id), Recruiter)
 
 
 @pytest.mark.django_db
 def test_get_recruiter_raises_exception_for_non_recruiter_user():
-    example_user_not_recruiter = User.objects.create_user("notrecruiter", "password")
+    example_user_not_recruiter = User.objects.create_user(
+        "notrecruiter", "password")
     example_user_not_recruiter.save()
     with pytest.raises(Recruiter.DoesNotExist):
         Recruiter.get_recruiter(example_user_not_recruiter.id)
+
+
+@pytest.mark.django_db
+def test_getting_the_correct_url(rf, admin_user):
+    request = rf.get('/recruiterRegister/')
+    # Remember that when using RequestFactory, the request does not pass
+    # through middleware. If your view expects fields such as request.user
+    # to be set, you need to set them explicitly.
+    # The following line sets request.user to an admin user.
+    request.user = admin_user
+    response = recruiterRegister(request)
+    assert response.status_code == 200
+
+
+@pytest.mark.django_db
+def test_sign_up_valid_recruiter(recruiter_example):
+    form = RecuiterRegistrationForm(data=recruiter_example)
+    if form.is_valid():
+        user = form.save()
+        assert User.objects.filter(pk=user.user.id).exists()
+        assert Recruiter.objects.filter(pk=user).exists()
+    else:
+        assert False
```

Comment:
If you get to this line, the test will fail because the form is invalid (probably because the data you pass to it from the fixture), and not because what you're actually testing which is the sign up process

Reply:
the forms can be invalid because of several reasons one of them is the fact that the user will send "bad" data. What I test here is that the Django form is actually working and save my data in the right places.  and that is an important part of the signup process no? @guy9050

Reply:
I don't think its useful to test the form object without testing the flow through the view - to test a form use `client` to submit data to the view and then check the stored data and returned values.

Reply:
I will add the client functionality to the test, but I think that a test the check if to form object himself do the right work and save to the correct places, is also relevant.

Reply:
This would be true only if the for testing would provide a significantly bigger amount of coverage, which it won't in this case since `test_new_recruiter_account_with_valid_data` tests exactly the same cod path. As it is its just more code you'd need to maintain that does not provide significant value.

Also having that 'if' and `assert false` makes the test look strange and hard to understand. You could have simply added `assert form.is_valid()` before the other assertions.

In general you want as little logic in your tests as possible in order to make then reliable. A good test has the following structure:

1. Setup of values and other input for the test (Could be partially/fully done in fixtures)
2. Call production code (Usually should be 1 line and avoid any complex logic)
3. Inspect results with assertions.

You should aim to not mix those parts as you do here - But again, this strange test is simply because of the somewhat awkward API of the form objects that were not really designed to be tested directly. The test that does this via the client is much more straight-forward.